### PR TITLE
Automatic Mixed Precision (AMP) support in wav2letter

### PR DESCRIPTION
**Pull request by mtmd**
Summary:
Implements Automatic Mixed Precision (AMP) support in wav2letter. Using the command line flag `--mixedprecision` it would be possible to enable mixed precision training for a certain model. The default value if `false`.

### CODE REVIEW

**Comment by mtmd** [comment]
@jacobkahn,  This should also resolve issue #375. I have the same problem when compiling with v.7.4.0.

**Comment by jacobkahn** [comment]
@mtmd fyi — I've fixed the linking issue you mentioned with respect to #375, so if you rebase on master, that change should go away.

File: Train.cpp [hunk 1]
```diff
@@ -537,69 +543,123 @@ int main(int argc, char** argv) {
...
+              crit->forward({output, fl::noGrad(sample[kTargetIdx])}).front();
+          af::sync();
+          if (FLAGS_mixedprecision) {
+            ++scaleCounter;
+            loss = loss * scaleFactor;
+          }
+          meters.fwdtimer.stopAndIncUnit();
+          meters.critfwdtimer.stopAndIncUnit();
+
+          if (af::anyTrue<bool>(af::isNaN(loss.array())) ||
```
Comment (jacobkahn) [inline Train.cpp]: This is an interesting idea for the scale factor and sample retrying, but I'm wondering — is this directly related to/required for AMP training in the experiments you did?

It's a significant change to the training flow, and I'd prefer if we moved the change to a separate PR so we can discuss in more detail if you were still able to train models with AMP without this scaling. That way we can be more deliberate about how best to integrate everything while ensuring good convergence.

File: src/common/Defines.cpp [hunk 2]
```diff
@@ -143,6 +143,15 @@ DEFINE_double(
     100,
     "percentage of training set (by number of utts) to use for evaluation");
 
+// Mixed Precision OPTIONS
+DEFINE_bool(mixedprecision, false, "Use mixed precision for training");
+DEFINE_int32(scalefactor, 4096, "Loss scale factor");
+DEFINE_int32(
+    scalefactorupdateinterval,
```
Comment (jacobkahn) [inline src/common/Defines.cpp]: See my comment above. If FP16 can work well-enough without this scaling, I'd prefer if we merge the needed changes as they are, then focus on changes needed to make those models converge better.

File: src/module/W2lModule.cpp [hunk 3]
```diff
@@ -417,6 +417,13 @@ std::shared_ptr<Module> parseLines(
         std::stoi(params[6]));
   }
 
+  /* ========== Precision Cast  ========== */
+  if (params[0] == "PC") {
+    LOG_IF(FATAL, params.size() != 2) << "Failed parsing - " << line;
+    auto targetType = fl::stringToAfType(params[1]);
```
Comment (jacobkahn) [inline src/module/W2lModule.cpp]: What's the intended behavior here if the user didn't specify `--mixedprecision` when they start training?

It might be reasonable to throw an exception if we're trying to add this module but the user has omitted the flag (or it wasn't included amongst the model's serialized flags). That way, the user can unambiguously know when they're dealing with an fp16 model and won't use fp16 components unless they explicitly want to or the model they're continuing/forking was already trained with them.

File: Train.cpp [hunk 4]
```diff
@@ -414,6 +414,8 @@ int main(int argc, char** argv) {
 
     for (auto& sample : *testds) {
       auto output = ntwrk->forward({fl::input(sample[kInputIdx])}).front();
+      output.cast(sample[kInputIdx].type());
+      af::sync();
```
[the remark this answers is not in the document]
Comment (mtmd) [reply Train.cpp]: fixed.

File: recipes/models/conv_glu/librispeech/network_amp.arch [hunk 5]
```diff
@@ -0,0 +1,58 @@
+V -1 1 NFEAT 0
```
[the remark this answers is not in the document]
Comment (mtmd) [reply recipes/models/conv_glu/librispeech/network_amp.arch]: Done.

File: Train.cpp (hunk 1 above)
Comment (mtmd) [reply Train.cpp]: Skipping them is a little bit inefficient computation-wise since at this point the sample is already loaded and preprocessed. In addition, we will be skipping samples that triggered a gradient explosion. That is, samples that the model needs to learn better.

File: src/module/W2lModule.cpp (hunk 3 above)
Comment (mtmd) [reply src/module/W2lModule.cpp]: Great Point. Fixed.

**Comment by jacobkahn** [comment]
@mtmd — as I mentioned in https://github.com/facebookresearch/flashlight/pull/72, flashlight and wav2letter now require ArrayFire >=, so we can begin merging this. Let me know when you've rebased on top of master/tested with a rebased flashlight PR and I'll do one final pass before we merge.

**Comment by joazoa** [comment]
I have been using this for a month now, it's working well. Does not go together with the transformer patch (some conflicts) and does not work with streaming_tds_model_converter.

**Comment by mtmd** [comment]
> @mtmd — as I mentioned in [facebookresearch/flashlight#72](https://github.com/facebookresearch/flashlight/pull/72), flashlight and wav2letter now require ArrayFire >=, so we can begin merging this. Let me know when you've rebased on top of master/tested with a rebased flashlight PR and I'll do one final pass before we merge.

@jacobkahn -- Rebased, tested, and pushed.
